cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,326 @@
+	ProvisioningClass string `json:"provisioningClass"`
+
+	// AdditionalParameters contains all other parameters custom classes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	AdditionalParameters map[string]string `json:"additionalParameters"`
+}
+
+type PodSet struct {
+	// PodTemplateRef is a reference to a PodTemplate object that is representing pods 
+	// that will consume this reservation (must be within the same namespace). 
+	// Users need to make sure that the  fields relevant to scheduler (e.g. node selector tolerations)
+	// are consistent between this template and actual pods consuming the Provisioning Request.
+	//
+	// +kubebuilder:validation:Required
+	PodTemplateRef Reference `json:"podTemplateRef"`
+	// Count contains the number of pods that will be created with a given
+	// template.
+	//
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=16384
+	Count int32 `json:"count"`
+}
+
+type Reference struct {
+	// Name of the referenced object.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name,omitempty"`
+}
+
+// ProvisioningRequestStatus represents the status of the resource reservation.
+type ProvisioningRequestStatus struct {
+	// Conditions represent the observations of a Provisioning Request's
+	// current state. Those will contain information whether the capacity
+    // was found/created or if there were any issues. The condition types
+    // may differ between different provisioning classes.
+	//
+	// +listType=map
+    // +listMapKey=type
+    // +patchStrategy=merge
+    // +patchMergeKey=type
+    // +optional
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// AdditionalStatus contains all other status values custom provisioning classes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=64
+	AdditionalStatus map[string]string `json:"additionalStatus"`
+}
+```
+
+### Provisioning Classes
+
+#### check-capacity.kubernetes.io class
+
+The `check-capacity.kubernetes.io` is one-off check to verify that the in the cluster
+there is enough capacity to provision given set of pods.
+
+Note: If two of such objects are created around the same time, CA will consider
+them independently and place no guards for the capacity.
+Also the capacity is not reserved in any manner so it may be scaled-down.
+
+#### atomic-scale-up.kubernetes.io class
+
+The `atomic-scale-up.kubernetes.io` aims to provision the resources required for the
+specified pods in an atomic way. The proposed logic is to:
+1. Try to provision required VMs in one loop.
+2. If it failed, remove the partially provisioned VMs and back-off.
+3. Stop the back-off after a given duration (optional), which would be passed
+   via `AdditionalParameters` field, using `ValidUntilSeconds` key and would contain string
```

Comment:
The field would be used only for one of the two modes that are proposed, so it would be skipped in the check-capacity mode.
Because of that it was moved to the map